### PR TITLE
Add support for generating ref-members in SyntaxGenerator

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/66376

### CODE REVIEW

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+*REMOVED*abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TypeExpression(Microsoft.CodeAnalysis.ITypeSymbol typeSymbol) -> Microsoft.CodeAnalysis.SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:1]: This is the change of an abstract method to a non-virtual one.  This is binary safe to make as this type is not publicly sublcassable (it has an internal constructor).

File: src/Workspaces/Core/Portable/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -139,7 +139,6 @@ abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TryCatchStatement(System
 abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TupleElementExpression(Microsoft.CodeAnalysis.SyntaxNode type, string name = null) -> Microsoft.CodeAnalysis.SyntaxNode
 abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TupleExpression(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> arguments) -> Microsoft.CodeAnalysis.SyntaxNode
 abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TypedConstantExpression(Microsoft.CodeAnalysis.TypedConstant value) -> Microsoft.CodeAnalysis.SyntaxNode
-abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.TypeExpression(Microsoft.CodeAnalysis.ITypeSymbol typeSymbol) -> Microsoft.CodeAnalysis.SyntaxNode
```
Comment (genlu) [inline src/Workspaces/Core/Portable/PublicAPI.Shipped.txt:142]: I might misremember things, but shouldn't we leave removed API in shipped.txt while adding a corresponding *REMOVED* entry in unshipped.txt until it ships?

**Review by genlu — APPROVED** [review]
Code change looks good to me

File: src/Workspaces/Core/Portable/PublicAPI.Shipped.txt (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/PublicAPI.Shipped.txt:142]: Oops. Yes.  I meant to copy, not cut here. Thanks!

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -3391,8 +3391,16 @@ internal override SyntaxNode GlobalAliasedName(SyntaxNode name)
...
-        public override SyntaxNode TypeExpression(ITypeSymbol typeSymbol)
-            => typeSymbol.GenerateTypeSyntax();
+        private protected override SyntaxNode TypeExpression(ITypeSymbol typeSymbol, RefKind refKind)
+        {
+            var type = typeSymbol.GenerateTypeSyntax();
+            return refKind switch
+            {
+                RefKind.Ref => SyntaxFactory.RefType(type),
+                RefKind.RefReadOnly => SyntaxFactory.RefType(SyntaxFactory.Token(SyntaxKind.RefKeyword), SyntaxFactory.Token(SyntaxKind.ReadOnlyKeyword), type),
+                _ => type,
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:3401]: ❔ Does this also need to handle parameters (where `out` and `in` would be modifiers)?

**Review by sharwell — APPROVED** [review]

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:3401]: Nope.  That is handled at hte ParmaeterDecl creation level itself :)

**Pull request merged**